src/components/dashboard/dashboard.service.ts

```diff
@@ -1090,6 +1090,9 @@ export class DashboardService implements OnDestroy {
         // check whether or not changes have been made - if so we need to repeat until stable
         let stable = true;
 
+        // update information about the layout
+        this.layout$.next(this.getLayoutData());
```

Comment:
This is being called before any layout changes are made. `shiftWidgetsUp` is also a recursive function, so this is potentially called multiple times. Is that intended?

`shiftWidgetsUp` is called in multiple contexts, not just as part of `refreshLayout`. Have you checked the other contexts as well to confirm that this is valid?

Perhaps an explanation of how this solves the problem in the PR description would be useful.

Reply:
Yeah good points here, I may be able to move update layout call out and into another function that is called before this.

I'll update the description as well. The problem being that the internal list of widgets isn't updated when one is removed. So when shiftWidgetsUp is called it still thinks that widget is still there and doesn't move it up.

But yeah this being recursive it wouldn't be ideal to have this called multiple times. I will have another play around with it.

Reply:
I'm not super familiar with the dashboard so hopefully this isn't misleading. But if the problem is the layout not updating after removing a widget, perhaps this could happen in the `removeWidget` function? That's called whenever a widget is destroyed.